_episodes/02-concepts.md

```diff
@@ -101,26 +145,40 @@ necessary computational environment for analysis.
 {: .callout}
 
 Alternatives that reproduce environments with minimal software 
-dependencies are technologies like virtual machines (virtualbox, vmwre, 
-NITRC CE), containers (docker, singularity) and installers (e.g., 
-vagrant, ansible). These can be very useful to replicate existing 
+dependencies are technologies like virtual machines (VirtualVox, VMWare, 
+NITRC CE), containers (Docker, Singularity) and installers (e.g., 
+Vagrant, Ansible). These can be very useful to replicate existing 
 environments and therefore simplify the installation problem 
 significantly. However, at present some of these technologies are not 
-installed by default on computing clusters you may have access to, 
-although that is changing.
+installed by default on computing clusters you may have access to.
 
 Lesson 3 covers container technologies and how to create, use, modify, 
 and reuse containers.
 
+> ## Exercise:
+>
+>  What should a reproducible computational environment contain?
+>
+> > ## Solution
+> >
+> > A reproducible computational environment must contain all the necessary data,
+> > environment variables, and software necessary for carrying out the ascribed 
+> > computation. Ideally, such an environment itself should be reproducible. 
+> {: .solution}
+{: .challenge}
+
+
+<a name="element3" />
+
 ### Element 3: Running analysis and storing expected results and provenance
 
 Once the environment is setup, one can execute the analysis. Each time
 the analysis is run the provenance of the workflow is captured and 
-stored using a PROV model for workflows. All of this happens inside a 
-single executable script. The script itself uses the Nipype dataflow 
-framework to ensure a consistent representation of the execution graph
-using Python as the dataflow language, and therefore benefits from all
-the advantages a dataflow framework brings to analyses.
+stored using a [PROV](https://www.w3.org/TR/prov-dm/) model for workflows. 
+All of this happens inside a single executable script. The script itself uses 
+the [Nipype dataflow framework](http://nipy.org/nipype) to ensure a consistent 
```

Comment:
It might be useful to add link the executable script : https://github.com/ReproNim/simple_workflow/blob/master/run_demo_workflow.py